File: bun/perfomance-tests/binary-search.ts
import { makeTetradicNumber, getLowestTetradicNumberIndex, makeTetradicNumberCalls } from './lib';
import { withTime } from './utils'


let count1 = 0
let count2 = 0
let count3 = 0
let count4 = 0
let count5 = 0

export function findSums(inputNumber: number): number[] | null {
  const nearestTetradicNumberIndex = getLowestTetradicNumberIndex(inputNumber);

  for (let number1Index = nearestTetradicNumberIndex; number1Index > 0; number1Index--) {
    const number1 = makeTetradicNumber(number1Index);
    count1++

    const required1 = inputNumber - number1;

    if (required1 < 0) {
      continue
    }

    if (number1 === inputNumber) {
      return [number1];
    }

    let left2 = 0;
    let right2 = number1Index + 1;
    let number2Index;

    while (left2 <= right2) {
      number2Index = Math.floor((left2 + right2) / 2);

      const number2 = makeTetradicNumber(number2Index);
      count2++

      if (number2 === required1) {
        return [number1, number2];
      }

      if (number2 < required1) {
        let left3 = 0;
        let right3 = number2Index + 1;
        let number3Index;

        const required2 = required1 - number2;

        if (required2 < 0) {
          continue
        }

        while (left3 <= right3) {
          number3Index = Math.floor((left3 + right3) / 2);

          const number3 = makeTetradicNumber(number3Index);
          count3++

          if (number3 === required2) {
            return [number1, number2, number3];
          }

          if (number3 < required2) {
            let left4 = 0;
            let right4 = number3Index + 1;
            let number4Index;

            const required3 = required2 - number3;

            if (required3 < 0) {
              continue
            }

            while (left4 <= right4) {
              number4Index = Math.floor((left4 + right4) / 2);

              const number4 = makeTetradicNumber(number4Index);
              count4++

              if (number4 === required3) {
                return [number1, number2, number3, number4];
              }

              if (number4 < required3) {
                let left5 = 0;
                let right5 = number4Index + 1;
                let number5Index;

                const required4 = required3 - number4;

                if (required4 < 0) {
                  continue
                }

                while (left5 <= right5) {
                  number5Index = Math.floor((left5 + right5) / 2);

                  const number5 = makeTetradicNumber(number5Index);
                  count5++

                  if (number5 === required4) {
                    return [number1, number2, number3, number4, number5];
                  }

                  if (number5 < required4) {
                    left5 = number5Index + 1;
                  } else {
                    right5 = number5Index - 1;
                  }
                }

                left4 = number4Index + 1;
              } else {
                right4 = number4Index - 1;
              }
            }

          }

          if (number3 < required2) {
            left3 = number3Index + 1;
          } else {
            right3 = number3Index - 1;
          }
        }


        left2 = number2Index + 1;
      } else {
        right2 = number2Index - 1;
      }
    }
  }

  return null;
}

const start = 999_999
const end = 1_000_000

const [result, time] = withTime(() => {
  for (let i = start; i < end; i++) {
    const result = findSums(i);
    if (i % 10000 === 0) {
      process.stdout.write(`\r${i} / ${end} (${Math.floor(i / end * 100)}%)`)
    }
    if (result === null || result.reduce((acc, curr) => acc + curr, 0) !== i || result.length > 5 || result.length < 0) {
      console.log('Failed at', i)
    }
  }
})

console.log(`Finished testing from ${start} to ${end} in ${time / 1000}s`)
console.log('makeTetradicNumber calls count', makeTetradicNumberCalls)
console.log('count1', count1)
console.log('count2', count2)
console.log('count3', count3)
console.log('count4', count4)
console.log('count5', count5)

File: tests/python/tests/binary-search.py
import sys

def make_tetradic_number(index):
  return (index * (index + 1) * (index + 2)) // 6

def make_tetradic_numbers(num):
  numbers = []
  index = 0

  while True:
    tetradic_number = make_tetradic_number(index)
    if tetradic_number > num:
      break

    numbers.append(tetradic_number)
    index += 1

  return numbers

def find_sums(input_number):
  tetradic_numbers = make_tetradic_numbers(input_number)
  length = len(tetradic_numbers)

  for number1_index in range(length - 1, 0, -1):
    number1 = tetradic_numbers[number1_index]
    required1 = input_number - number1

    if number1 == input_number:
      return [number1]

    left2, right2 = 0, number1_index + 1

    while left2 <= right2:
      number2_index = (left2 + right2) // 2
      number2 = tetradic_numbers[number2_index]

      if number2 == required1:
        return [number1, number2]

      if number2 < required1:
        required2 = required1 - number2

        if required2 < 0:
          continue

        left3, right3 = 0, number2_index + 1

        while left3 <= right3:
          number3_index = (left3 + right3) // 2
          number3 = tetradic_numbers[number3_index]

          if number3 == required2:
            return [number1, number2, number3]

          if number3 < required2:
            required3 = required2 - number3

            if required3 < 0:
              continue

            left4, right4 = 0, number3_index + 1

            while left4 <= right4:
              number4_index = (left4 + right4) // 2
              number4 = tetradic_numbers[number4_index]

              if number4 == required3:
                return [number1, number2, number3, number4]

              if number4 < required3:
                required4 = required3 - number4

                if required4 < 0:
                  continue

                left5, right5 = 0, number4_index + 1

                while left5 <= right5:
                  number5_index = (left5 + right5) // 2
                  number5 = tetradic_numbers[number5_index]

                  if number5 == required4:
                    return [number1, number2, number3, number4, number5]

                  if number5 < required4:
                    left5 = number5_index + 1
                  else:
                    right5 = number5_index - 1

                left4 = number4_index + 1
              else:
                right4 = number4_index - 1

          if number3 < required2:
            left3 = number3_index + 1
          else:
            right3 = number3_index - 1

        left2 = number2_index + 1
      else:
        right2 = number2_index - 1

  return None

def test(start, end, step):
  for i in range(start, end):
    result = find_sums(i)

    if result is None or sum(result) != i:
      raise ValueError(f"No result for {i}")

if __name__ == "__main__":
  args = list(map(int, sys.argv[1:4]))
  test(*args)